File: scff_app/scff_interprocess/interprocess.cs
// Copyright 2012 Alalf <alalf.iQLc_at_gmail.com>
//
// This file is part of SCFF-DirectShow-Filter(SCFF DSF).
//
// SCFF DSF is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// SCFF DSF is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with SCFF DSF.  If not, see <http://www.gnu.org/licenses/>.

/// @file scff_interprocess/interprocess.cs
/// SCFFのプロセス間通信に関する定数、型の宣言
/// @warning To me: このファイルの中から別のファイルへのusingは禁止！
/// - 別の言語に移植する場合も最大2ファイルでお願いします

/// scff_interprocessモジュールのC#版(オリジナルはC++)
namespace scff_interprocess {

using System;
using System.Runtime.InteropServices;

//=====================================================================
/// @page smp SCFF Messaging Protocol v1 (by 2012/05/22 Alalf)
/// SCFF_DSFおよびそのクライアントで共有する共有メモリ内のデータ配置の仕様
///
/// [全体的な注意点]
/// - Windows固有の型名はビットサイズが分かりにくいのでSystem.***で置き換える
///   - 対応表
///     - DWORD        = UInt32 (32bit)
///     - HWND(void*)  = UInt64 (64bit)
///       - SCFHから変更: 念のため32bitから64bitに
///     - bool         = Byte (8bit)
/// - 不動小数点数は基本的にはdoubleで(floatも利用可)
///   - 対応表
///     - double       = Double (64bit)
///     - float        = Single (32bit)
/// - すべての構造体はPOD(Plain Old Data)であること
///   - 基本型、コンストラクタ、デストラクタ、仮想関数を持たない構造体のみ
//=====================================================================

/// プロセス間通信を担当するクラス
partial class Interprocess {
  /// 共有メモリ名: SCFFエントリを格納するディレクトリ
  const string kDirectoryName = "scff_v1_directory";

  /// Directoryの保護用Mutex名
  const string kDirectoryMutexName = "mutex_scff_v1_directory";

  /// 共有メモリ名の接頭辞: SCFFで使うメッセージを格納する
  const string kMessageNamePrefix = "scff_v1_message_";

  /// Messageの保護用Mutex名の接頭辞
  const string kMessageMutexNamePrefix = "mutex_scff_v1_message_";

  //-------------------------------------------------------------------

  /// Path文字列の長さ
  public const int kMaxPath = 260;

  /// Directoryに格納されるEntryの最大の数
  public const int kMaxEntry = 8;

  /// ComplexLayout利用時の最大の要素数
  /// @sa scff_imaging::kMaxProcessorSize
  public const int kMaxComplexLayoutElements = 8;
}

//-------------------------------------------------------------------

/// レイアウトの種類
public enum LayoutTypes {
  /// 何も表示しない
  kNullLayout = 0,
  /// 取り込み範囲1個で、境界は出力に強制的に合わせられる
  kNativeLayout,
  /// 取り込み範囲が複数ある
  kComplexLayout
}

//-------------------------------------------------------------------

/// ピクセルフォーマットの種類
/// @sa scff_imaging/imaging_types.h
/// @sa scff_imaging::ImagePixelFormats
public enum ImagePixelFormats {
  /// 不正なピクセルフォーマット
  kInvalidPixelFormat = -1,
  /// I420(12bit)
  kI420 = 0,
  /// IYUV(12bit)
  kIYUV,
  /// YV12(12bit)
  kYV12,
  /// UYVY(16bit)
  kUYVY,
  /// YUY2(16bit)
  kYUY2,
  /// RGB0(32bit)
  kRGB0,
  /// 対応ピクセルフォーマット数
  kSupportedPixelFormatsCount
}

//-------------------------------------------------------------------

/// 拡大縮小メソッドをあらわす定数
/// @sa scff_imaging/imaging_types.h
/// @sa scff_imaging::SWScaleFlags
public enum SWScaleFlags {
  /// fast bilinear
  kFastBilinear = 1,
  /// bilinear
  kBilinear = 2,
  /// bicubic
  kBicubic = 4,
  /// experimental
  kX = 8,
  /// nearest neighbor
  kPoint = 0x10,
  /// averaging area
  kArea = 0x20,
  /// luma bicubic, chroma bilinear
  kBicublin = 0x40,
  /// gaussian
  kGauss = 0x80,
  /// sinc
  kSinc = 0x100,
  /// lanczos
  kLanczos = 0x200,
  /// natural bicubic spline
  kSpline = 0x400
}

/// 回転方向を表す定数
/// @sa scff_imaging/imaging_types.h
/// @sa scff_imaging::RotateDirections
public enum RotateDirections {
  /// 回転なし
  kNoRotate = 0,
  /// 時計回り90度
  kDegrees90,
  /// 時計回り180度
  kDegrees180,
  /// 時計回り270度
  kDegrees270
}

/// 共有メモリ(Directory)に格納する構造体のエントリ
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct Entry {
  /// SCFF DSFのDLLが使われれているプロセスID
  public UInt32 process_id;
  /// SCFF DSFのDLLが使われているプロセス名
  [MarshalAs(UnmanagedType.ByValTStr, SizeConst = Interprocess.kMaxPath)]
  public string process_name;
  /// サンプルの出力width
  public Int32 sample_width;
  /// サンプルの出力height
  public Int32 sample_height;
  /// サンプルの出力ピクセルフォーマット
  /// @attention ImagePixelFormatを操作に使うこと
  public Int32 sample_pixel_format;
  /// 目標fps
  public Double fps;
}

/// 共有メモリ(Directory)に格納する構造体
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct Directory {
  [MarshalAs(UnmanagedType.ByValArray, SizeConst = Interprocess.kMaxEntry)]
  public Entry[] entries;
}

/// 拡大縮小設定
/// @sa scff_imaging::SWScaleConfig
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct SWScaleConfig {
  /// 拡大縮小メソッド(Chroma/Luma共通)
  /// @attention 操作にはSWScaleFlagsを使うこと！
  public Int32 flags;
  /// 正確な丸め処理
  public Byte accurate_rnd;
  /// 変換前にフィルタをかけるか
  public Byte is_filter_enabled;
  /// 輝度のガウスぼかし
  public Single luma_gblur;
  /// 色差のガウスぼかし
  public Single chroma_gblur;
  /// 輝度のシャープ化
  public Single luma_sharpen;
  /// 色差のシャープ化
  public Single chroma_sharpen;
  /// 水平方向のワープ
  public Single chroma_hshift;
  /// 垂直方向のワープ
  public Single chroma_vshift;
};

/// レイアウトパラメータ
/// @sa scff_imaging::ScreenCaptureParameter
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct LayoutParameter {
  /// サンプル内の原点のX座標
  /// @warning NullLayout,NativeLayoutでは無視される
  public Int32 bound_x;
  /// サンプル内の原点のY座標
  /// @warning NullLayout,NativeLayoutでは無視される
  public Int32 bound_y;
  /// サンプル内の幅
  /// @warning NullLayout,NativeLayoutでは無視される
  public Int32 bound_width;
  /// サンプル内の高さ
  /// @warning NullLayout,NativeLayoutでは無視される
  public Int32 bound_height;
  /// キャプチャを行う対象となるウィンドウ
  public UInt64 window;
  /// 取り込み範囲の開始X座標
  public Int32 clipping_x;
  /// 取り込み範囲の開始y座標
  public Int32 clipping_y;
  /// 取り込み範囲の幅
  public Int32 clipping_width;
  /// 取り込み範囲の高さ
  public Int32 clipping_height;
  /// マウスカーソルの表示
  public Byte show_cursor;
  /// レイヤードウィンドウの表示
  public Byte show_layered_window;
  /// 拡大縮小設定
  public SWScaleConfig swscale_config;
  /// 取り込み範囲が出力サイズより小さい場合拡張
  public Byte stretch;
  /// アスペクト比の保持
  public Byte keep_aspect_ratio;
  /// 回転方向
  /// @attention RotateDirectionを操作に使うこと
  public Int32 rotate_direction;
}

/// 共有メモリ(Message)に格納する構造体
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct Message {
  /// タイムスタンプ(time()で求められたものを想定)
  /// @warning 1ではじまり単調増加が必須条件
  /// @warning (0および負数は無効なメッセージを示す)
  public Int64 timestamp;
  /// レイアウトの種類
  /// @attention LayoutTypeを操作に使うこと
  public Int32 layout_type;
  /// 有効なレイアウト要素の数
  public Int32 layout_element_count;
  /// レイアウトパラメータの配列
  [MarshalAs(UnmanagedType.ByValArray, SizeConst = Interprocess.kMaxComplexLayoutElements)]
  public LayoutParameter[] layout_parameters;
}
}

File: scff_dsf/scff_interprocess/interprocess.cc
// Copyright 2012-2013 Alalf <alalf.iQLc_at_gmail.com>
//
// This file is part of SCFF-DirectShow-Filter(SCFF DSF).
//
// SCFF DSF is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// SCFF DSF is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with SCFF DSF.  If not, see <http://www.gnu.org/licenses/>.

/// @file scff_interprocess/interprocess.cc
/// SCFFのプロセス間通信に関するクラスの定義
/// @warning To me: このファイルの中から別のファイルへのIncludeは禁止！
/// - 別の言語に移植する場合も最大2ファイルでお願いします

#include "scff_interprocess/interprocess.h"

#include <stdio.h>
#include <tchar.h>

namespace scff_interprocess {

//=====================================================================
// scff_interprocess::Interprocess
//=====================================================================

Interprocess::Interprocess()
    : directory_(nullptr),
      view_of_directory_(nullptr),
      mutex_directory_(nullptr),
      message_(nullptr),
      view_of_message_(nullptr),
      mutex_message_(nullptr),
      error_event_(nullptr),
      shutdown_event_(nullptr) {
  // nop
  OutputDebugString(TEXT("****Interprocess: NEW\n"));
}

Interprocess::~Interprocess() {
  OutputDebugString(TEXT("****Interprocess: DELETE\n"));
  // 解放忘れがないように
  ReleaseErrorEvent();
  ReleaseShutdownEvent();
  ReleaseMessage();
  ReleaseDirectory();
}

//---------------------------------------------------------------------

bool Interprocess::InitDirectory() {
  // 念のため解放
  ReleaseDirectory();

  // 仮想メモリ(Directory)の作成
  HANDLE tmp_directory =
      CreateFileMappingA(INVALID_HANDLE_VALUE,
                          nullptr,
                          PAGE_READWRITE,
                          0,
                          sizeof(Directory),
                          kDirectoryName);
  if (tmp_directory == nullptr) {
    // 仮想メモリ作成失敗
    return false;
  }
  DWORD error_create_file_mapping = GetLastError();

  // ビューの作成
  LPVOID tmp_view_of_directory =
      MapViewOfFile(tmp_directory,
                    FILE_MAP_ALL_ACCESS,
                    0, 0, 0);
  if (tmp_view_of_directory == nullptr) {
    // ビュー作成失敗
    CloseHandle(tmp_directory);
    return false;
  }

  // 最初に共有メモリを作成した場合は0クリアしておく
  if (tmp_view_of_directory != nullptr &&
      error_create_file_mapping != ERROR_ALREADY_EXISTS) {
    ZeroMemory(tmp_view_of_directory, sizeof(Directory));
  }

  // Mutexの作成
  HANDLE tmp_mutex_directory =
      CreateMutexA(nullptr, false, kDirectoryMutexName);
  if (tmp_mutex_directory == nullptr) {
    // Mutex作成失敗
    UnmapViewOfFile(tmp_view_of_directory);
    CloseHandle(tmp_directory);
    return false;
  }
  DWORD error_create_mutex = GetLastError();

  // 最初にMutexを作成した場合は…なにもしなくていい
  if (tmp_mutex_directory != nullptr &&
      error_create_mutex != ERROR_ALREADY_EXISTS) {
    // nop
  }

  // メンバ変数に設定
  directory_ = tmp_directory;
  view_of_directory_ = tmp_view_of_directory;
  mutex_directory_ = tmp_mutex_directory;

  OutputDebugString(TEXT("****Interprocess: InitDirectory Done\n"));
  return true;
}

bool Interprocess::IsDirectoryInitialized() {
  return directory_ != nullptr &&
         view_of_directory_ != nullptr &&
         mutex_directory_ != nullptr;
}

void Interprocess::ReleaseDirectory() {
  if (mutex_directory_ != nullptr) {
    CloseHandle(mutex_directory_);
    mutex_directory_ = nullptr;
  }
  if (view_of_directory_ != nullptr) {
    UnmapViewOfFile(view_of_directory_);
    view_of_directory_ = nullptr;
  }
  if (directory_ != nullptr) {
    CloseHandle(directory_);
    directory_ = nullptr;
  }
}

//---------------------------------------------------------------------

bool Interprocess::InitMessage(uint32_t process_id) {
  // 念のため解放
  ReleaseMessage();

  // 仮想メモリの名前
  char message_name[256];
  ZeroMemory(message_name, sizeof(message_name));
  sprintf_s(message_name,
            256, "%s%d",
            kMessageNamePrefix, process_id);

  // 仮想メモリ(Message<process_id>)の作成
  HANDLE tmp_message =
      CreateFileMappingA(INVALID_HANDLE_VALUE,
                          nullptr,
                          PAGE_READWRITE,
                          0,
                          sizeof(Message),
                          message_name);
  if (tmp_message == nullptr) {
    // 仮想メモリ作成失敗
    return false;
  }
  DWORD error_create_file_mapping = GetLastError();

  // ビューの作成
  LPVOID tmp_view_of_message =
      MapViewOfFile(tmp_message,
                    FILE_MAP_ALL_ACCESS,
                    0, 0, 0);
  if (tmp_view_of_message == nullptr) {
    // ビュー作成失敗
    CloseHandle(tmp_message);
    return false;
  }

  // 最初に共有メモリを作成した場合は0クリアしておく
  if (tmp_view_of_message != nullptr &&
      error_create_file_mapping != ERROR_ALREADY_EXISTS) {
    ZeroMemory(tmp_view_of_message, sizeof(Message));
  }

  // Mutexの名前
  char message_mutex_name[256];
  ZeroMemory(message_mutex_name, sizeof(message_mutex_name));
  sprintf_s(message_mutex_name,
            256, "%s%d",
            kMessageMutexNamePrefix, process_id);

  // Mutexの作成
  HANDLE tmp_mutex_message =
      CreateMutexA(nullptr, false, message_mutex_name);
  if (tmp_mutex_message == nullptr) {
    // Mutex作成失敗
    UnmapViewOfFile(tmp_view_of_message);
    CloseHandle(tmp_message);
    return false;
  }
  DWORD error_create_mutex = GetLastError();

  // 最初にMutexを作成した場合は…なにもしなくていい
  if (tmp_mutex_message != nullptr &&
      error_create_mutex != ERROR_ALREADY_EXISTS) {
    // nop
  }

  // メンバ変数に設定
  message_ = tmp_message;
  view_of_message_ = tmp_view_of_message;
  mutex_message_ = tmp_mutex_message;

  OutputDebugString(TEXT("****Interprocess: InitMessage Done\n"));
  return true;
}

bool Interprocess::IsMessageInitialized() {
  return message_ != nullptr &&
         view_of_message_ != nullptr &&
         mutex_message_ != nullptr;
}

void Interprocess::ReleaseMessage() {
  if (mutex_message_ != nullptr) {
    CloseHandle(mutex_message_);
    mutex_message_ = nullptr;
  }
  if (view_of_message_ != nullptr) {
    UnmapViewOfFile(view_of_message_);
    view_of_message_ = nullptr;
  }
  if (message_ != nullptr) {
    CloseHandle(message_);
    message_ = nullptr;
  }
}

//---------------------------------------------------------------------

bool Interprocess::InitErrorEvent(uint32_t process_id) {
  // 念のため解放
  ReleaseErrorEvent();

  // イベントの名前
  TCHAR error_event_name[256];
  ZeroMemory(error_event_name, sizeof(error_event_name));
  _stprintf_s(error_event_name,
              256, TEXT("%s%d"),
              kErrorEventNamePrefix, process_id);
            
  // イベント(ErrorEvent<process_id>)の作成
  HANDLE tmp_error_event = CreateEvent(nullptr, FALSE, FALSE, error_event_name);
  if (tmp_error_event == nullptr) {
    // イベント作成失敗
    return false;
  }
  DWORD error_create_event = GetLastError();

  // 最初にEventを作成した場合は…なにもしなくていい
  if (error_create_event != ERROR_ALREADY_EXISTS) {
    // nop
  }

  // メンバ変数に設定
  error_event_ = tmp_error_event;

  OutputDebugString(TEXT("****Interprocess: InitErrorEvent Done\n"));
  return true;
}

bool Interprocess::IsErrorEventInitialized() {
  return error_event_ != nullptr;
}

void Interprocess::ReleaseErrorEvent() {
  if (error_event_ != nullptr) {
    CloseHandle(error_event_);
    error_event_ = nullptr;
  }
}

//---------------------------------------------------------------------

bool Interprocess::InitShutdownEvent() {
  // プログラム全体で一回だけCreate/CloseすればよいのでCloseはしない
  if (IsShutdownEventInitialized()) return true;
            
  // イベント(ShutdownEvent)の作成
  HANDLE tmp_shutdown_event = CreateEvent(nullptr, TRUE, FALSE, nullptr);
  if (tmp_shutdown_event == nullptr) {
    // イベント作成失敗
    return false;
  }

  // メンバ変数に設定
  shutdown_event_ = tmp_shutdown_event;

  OutputDebugString(TEXT("****Interprocess: InitShutdownEvent Done\n"));
  return true;
}

bool Interprocess::IsShutdownEventInitialized() {
  return shutdown_event_ != nullptr;
}

void Interprocess::ReleaseShutdownEvent() {
  if (shutdown_event_ != nullptr) {
    CloseHandle(shutdown_event_);
    shutdown_event_ = nullptr;
  }
}

//-------------------------------------------------------------------
// for SCFF DirectShow Filter
//-------------------------------------------------------------------

bool Interprocess::AddEntry(const Entry &entry) {
  // 初期化されていなければ失敗
  if (!IsDirectoryInitialized()) {
    OutputDebugString(TEXT("****Interprocess: AddEntry FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: AddEntry\n"));

  // ロック取得
  WaitForSingleObject(mutex_directory_, INFINITE);

  Directory *directory =
      static_cast<Directory*>(view_of_directory_);
  bool success = false;
  for (int i = 0; i < kMaxEntry; i++) {
    /// @warning 重複する場合がある？
    if (directory->entries[i].process_id == 0) {
      // PODなのでコピー可能（のはず）
      directory->entries[i] = entry;
      success = true;
      break;
    }
  }

  // ロック解放
  ReleaseMutex(mutex_directory_);

  if (success) {
    OutputDebugString(TEXT("****Interprocess: AddEntry SUCCESS\n"));
  }

  return success;
}

bool Interprocess::RemoveEntry(uint32_t process_id) {
  // 初期化されていなければ失敗
  if (!IsDirectoryInitialized()) {
    OutputDebugString(TEXT("****Interprocess: RemoveEntry FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: RemoveEntry\n"));

  // ロック取得
  WaitForSingleObject(mutex_directory_, INFINITE);

  Directory *directory =
      static_cast<Directory*>(view_of_directory_);
  for (int i = 0; i < kMaxEntry; i++) {
    if (directory->entries[i].process_id == process_id) {
      ZeroMemory(&(directory->entries[i]), sizeof(directory->entries[i]));
      OutputDebugString(TEXT("****Interprocess: RemoveEntry SUCCESS\n"));
      break;
    }
  }

  // ロック解放
  ReleaseMutex(mutex_directory_);

  return true;
}

bool Interprocess::ReceiveMessage(Message *message) {
  // 初期化されていなければ失敗
  if (!IsMessageInitialized()) {
    OutputDebugString(TEXT("****Interprocess: ReceiveMessage FAILED\n"));
    return false;
  }

  //OutputDebugString(TEXT("****Interprocess: ReceiveMessage\n"));

  // ロック取得
  WaitForSingleObject(mutex_message_, INFINITE);

  // そのままコピー
  memcpy(message, view_of_message_, sizeof(Message));

  // ロック解放
  ReleaseMutex(mutex_message_);

  return true;
}

bool Interprocess::RaiseErrorEvent() {
  // 初期化されていなければ失敗
  if (!IsErrorEventInitialized()) {
    OutputDebugString(TEXT("****Interprocess: RaiseErrorEvent FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: RaiseErrorEvent\n"));

  // シグナルを送る
  BOOL error_set_event = SetEvent(error_event_);
  if (!error_set_event) {
    OutputDebugString(TEXT("****Interprocess: RaiseErrorEvent FAILED\n"));
    return false;
  }

  return true;
}

//-------------------------------------------------------------------

bool Interprocess::GetDirectory(Directory *directory) {
  // 初期化されていなければ失敗
  if (!IsDirectoryInitialized()) {
    OutputDebugString(TEXT("****Interprocess: GetDirectory FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: GetDirectory\n"));

  // ロック取得
  WaitForSingleObject(mutex_directory_, INFINITE);

  // そのままコピー
  memcpy(directory, view_of_directory_, sizeof(Directory));

  // ロック解放
  ReleaseMutex(mutex_directory_);

  return true;
}

bool Interprocess::SendMessage(const Message &message) {
  // 初期化されていなければ失敗
  if (!IsMessageInitialized()) {
    OutputDebugString(TEXT("****Interprocess: SendMessage FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: SendMessage\n"));

  // ロック取得
  WaitForSingleObject(mutex_message_, INFINITE);

  // そのままコピー
  memcpy(view_of_message_, &message, sizeof(Message));

  // ロック解放
  ReleaseMutex(mutex_message_);

  return true;
}

bool Interprocess::WaitUntilErrorEventOccured() {
  // 初期化されていなければ失敗
  if (!IsErrorEventInitialized()) {
    OutputDebugString(TEXT("****Interprocess: WaitUntilErrorEventOccured FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: WaitUntilErrorEventOccured\n"));

  // シグナル状態になるまで待機
  const HANDLE events[] = {error_event_, shutdown_event_};
  const int signaled_event_index = WaitForMultipleObjects(2, events, FALSE, INFINITE);
  if (signaled_event_index != 0) {
    // エラー以外のイベントが起きた場合は失敗
    OutputDebugString(TEXT("****Interprocess: WaitUntilErrorEventOccured FAILED\n"));
    return false;
  }

  return true;
}

bool Interprocess::RaiseShutdownEvent() {
  // 初期化されていなければ失敗
  if (!IsShutdownEventInitialized()) {
    OutputDebugString(TEXT("****Interprocess: RaiseShutdownEvent FAILED\n"));
    return false;
  }

  OutputDebugString(TEXT("****Interprocess: RaiseShutdownEvent\n"));

  // シグナルを送る
  BOOL error_set_event = SetEvent(shutdown_event_);
  if (!error_set_event) {
    OutputDebugString(TEXT("****Interprocess: RaiseShutdownEvent FAILED\n"));
    return false;
  }

  return true;
}
}   // namespace scff_interprocess
